### ISSUE TITLE
Refuse a training run that cannot converge: a step-count guard between prepare and train

### ISSUE THREAD

**Issue by thodel** [thodel/serving-atr-inference#72]
The remaining half of #52, and the guard that would have saved the two kraken runs.

## What happened without it

`kraken-thun-missiven-v1` was submitted with the documented defaults and completed
honestly, reporting **CER 0.9838**. The arithmetic that explains it was available
the moment `prepare` finished:

| | |
|---|---|
| lines materialized | 2,087 |
| training lines (`partition: 0.9`) | ~1,878 |
| `batch_size` | 256 |
| steps per epoch | **7** |
| `epochs` | 50 |
| total optimizer steps | **~367** |
| parameters, from random init | 15.2 M |

367 steps for a 15.2 M-parameter network from scratch, with `1cycle` ramping and
annealing the learning rate across all of them, cannot converge. An unconverged
CTC network has not learned blank-dominance and emits a character at nearly every
timestep — which *is* the insertion-dominated CER that was then spent two days
diagnosing (11,191 insertions against 2 deletions).

Nothing between "submit" and "completed" asked whether the configuration could
learn anything. Every guard in the subsystem protects the disk, the GPU, the
filesystem or the reporting; none protects the **experiment**.

## Where it goes

Not at submit, despite the shorthand: the line count is not known until `prepare`
has materialized the pages. The natural point is **between `prepare` and `train`**,
where `job.progress.lines_written` has just been written and no GPU time has been
spent yet. `KrakenTrainParams.effective_batch_size` (`batch_size ×
accumulate_grad_batches`) already exists on all three params models, so the
arithmetic is:

```python
steps_per_epoch = ceil(train_lines / params.effective_batch_size)
total_steps     = steps_per_epoch * params.epochs
```

A cheaper pre-check at submit is possible later if row counts can be read from the
parquet metadata during `verify_dataset_spec` (#46) — worth doing, but it does not
block this.

## What it should do

**Refuse, not warn.** A warning in a log nobody reads is what the current state
already amounts to. A run that cannot converge should not consume a GPU for three
hours — the job fails at the boundary with the arithmetic in `job.error`, the same
way a full disk or a missing venv is refused.

The message has to be the whole story, in the shape the other guards use:

> 1,878 training lines at effective batch 256 is 7 steps/epoch; over 50 epochs
> that is 367 optimizer steps for a 15.2 M-parameter network from scratch, which
> will not converge. Either fine-tune from a base model (`base_model` + `resize:
> union`), or lower `batch_size` to ~16 (≈5,900 steps), or raise `epochs`.

**Engine-aware thresholds.** A QLoRA fine-tune of an 8 B VLM converges in tens of
steps — `qwen3vl-thun-smoke` was 38 steps and improved CER from 1.837 to 0.466 —
while a from-scratch CTC network needs thousands. One number cannot serve both. At
minimum: a per-engine floor, and a *separate, higher* floor when `base_model` is
None (training from scratch), since that is the case that failed.

**Overridable, and recorded when overridden.** Someone deliberately running a
90-second smoke test must be able to; `force: true` on the request, with the fact
that the guard was overridden on the job record so the resulting CER is never read
as an ordinary one.

## Acceptance

- A job whose configuration yields too few steps fails at the prepare→train
  boundary, before any GPU work, with the arithmetic and both remedies in the error.
- Replaying the three historical runs' recorded numbers through the guard flags
  `kraken-thun-missiven-v1` and passes `qwen3vl-thun-smoke` — the fixtures are in
  `docs/TRAINING_PLAN.md` §9a.
- `accumulate_grad_batches` and `partition` are honoured in the arithmetic (a run
  at `batch_size: 64, accumulate_grad_batches: 4` has the same effective batch as
  256 and must be judged the same).
- The override is respected and visible on the record.
- `docs/TRAINING.md` §"Do not copy the defaults onto a small corpus" points at the
  guard rather than only warning in prose.

Part of #49. The diagnosis it comes from is 1d18f2a; the material audit that ruled
out the other explanation is `scripts/audit_eval_material.py`.

**Comment by thodel** [comment]
Landed in ac7a94f. CI green, 602 tests.

**Where it ended up:** after `prepare`, before `compile` — the issue said "before train", but compile costs real time and produces nothing worth having if the run is doomed. Prepare is also the first moment a line count exists, which is why this cannot live at submit; a genuine submit-time check needs row counts from parquet metadata, which stays noted against #46.

**The refusal carries the whole story** rather than a verdict: line count, effective batch, steps/epoch, total steps, the floor, and three ways out — fine-tune from a base model, lower `batch_size` to a named value that actually clears the floor, or raise `epochs`. There is a test asserting the suggested batch really does clear it, because a remedy that is merely a smaller number is not a remedy.

**Floors** are in `convergence.py` at the top, not buried: 2,000 steps from scratch, 500 for a fine-tune, 25 for `vllm`. The last one is set below 38 deliberately — `qwen3vl-thun-smoke` improved CER 1.837 → 0.466 in 38 steps, and a floor in the hundreds would refuse the one run on this box that has demonstrably worked. They are judgement calls, not measurements, and the docstring says so.

**Acceptance, item by item:**
- Refused at the boundary before GPU work, arithmetic and both remedies in `job.error` — yes, and a test asserts `compile` and `train` are never invoked.
- Historical runs as fixtures — `kraken-thun-missiven-v1` (1,878 lines, batch 256, 50 epochs) is refused; `qwen3vl-thun-smoke` passes.
- `accumulate_grad_batches` and `partition` honoured — batch 64 × accumulate 4 is judged identically to batch 256, since that substitution is the runbook's own OOM remedy and must not walk past the guard. `train_lines` is now tracked separately from `lines_written`, which includes the eval side; dividing by the total would flatter every configuration.
- Override respected and visible — `"force": true` runs it and writes `convergence_override` onto the record.
- `docs/TRAINING.md` updated where the warning already was.

One consequence worth flagging: the pipeline fixtures now pass `force=True`. They push two or three fake pages through the whole lifecycle, which is far below anything the guard should allow — they are about the stages, so they opt out rather than pretending to be real runs.

Not exercised against a real job yet. The first real submission after this is the test: a from-scratch job on a small corpus should now fail in seconds with the arithmetic, where it used to fail in three hours with a CER.
